Add jira trackers

costaconrado opened the pull request:
Add a new module for filling trackers.

This PR introduce a new module capable of filling Jira trackers.
This PR also creates a new collector and new model to store Jira projects metadata, excluding some projects that are marked in products definition as Jira tracked but for some reason does not contains a valid Jira project (e.g `PSSVCS`, `RHBRMS`, `SSP` and others).

Closes OSIDB-89.

osoukup reviewed (commented):
I reviewed only a very little part and had to switch to other task. I will continue next time.

osoukup commented on apps/trackers/service.py:
```diff
@@ -0,0 +1,105 @@
+    """
+    Jira query handler for tracker management.
+    This class encapsulates calls for Jira doing validations
+    and it methods return data requested with HTTP status code
```
nitpick: its methods

costaconrado replied:
Fixed, thank you.

osoukup commented on apps/trackers/service.py:
```diff
@@ -0,0 +1,105 @@
+        If multiples streams is given creates one tracker per stream.
+        """
+        if affect.trackers.exists():
+            return Response(data="Affect already have trackers created.", status=409)
```
I am not sure that this check is correct. We can file different trackers for a specific affect in different times. There is no existing restriction on this. Having some trackers already should not be a blocker for creating new ones.

Of course there should be only a single tracker for each stream so we should restrict this.

costaconrado replied:
changed validation during the tracker creation in order to limit 1 tracker per stream for each affect.

osoukup commented on apps/trackers/service.py:
```diff
@@ -0,0 +1,105 @@
+
+        # creates a list of trackers to be generated in BTS
+        bts_trackers = []
+        if not tracked_streams:
```
The stream is a vital part of the tracker so I do not think it should ever be empty.

costaconrado replied:
Makes sense, enforced the stream.

osoukup commented on apps/trackers/service.py:
```diff
@@ -0,0 +1,105 @@
+                bulk_issues.append(bts_tracker.generate_bts_object()["fields"])
+
+            response = self.jira_conn.create_issues(bulk_issues, prefetch=True)
+            issues = response
```
What is the reason for this line? Why do we need two variables? Without copy it is just two references to the same thing anyway. Or maybe I miss something :smile:

costaconrado replied:
We don't need it, I missed it when refactoring this piece of code :sweat_smile:

osoukup commented on apps/trackers/service.py:
```diff
@@ -0,0 +1,105 @@
+                while i < len(response):
+                    issue = issues[i].raw
+                    bts_tracker = bts_trackers[i]
+                    stream = "" if not bts_tracker._stream else bts_tracker._stream.name
```
We should not permit creating new trackers without the stream. There can be some historical or invalid missing it but the newly created should be all valid.

costaconrado replied:
Makes sense, enforced the stream.

costaconrado commented on apps/trackers/apps.py:
```diff
@@ -0,0 +1,9 @@
+"""
+Trackers - everything related to the bug trackers.
```
makes sense, changed for vulnerability.

costaconrado commented on collectors/jiraffe/tasks.py:
```diff
@@ -21,3 +21,17 @@
+
+@collector(
+    base=MetadataCollector,
+    # run once a day at 3:03
```
Makes sense, I have changed for a time that has no other collectors running.

costaconrado commented on collectors/jiraffe/tests/test_collectors.py:
```diff
@@ -149,3 +156,27 @@ def test_collect_id_known_linked(self):
+    @pytest.mark.vcr
+    def test_collect(self, monkeypatch):
+        """
+        Test that collector are able ot get metadata from Jira projects
```
Fixed, thank you.

costaconrado commented on apps/trackers/bts_tracker.py:
```diff
@@ -0,0 +1,157 @@
+        }
+        if self._flaw.is_embargoed:
+            bts_tracker["fields"]["security"] = {"name": "Embargoed Security Issue"}
+        stream_name = "" if not self._stream else self._stream.name
```
I'm not sure if it would have any usage, I only added the empty possibility because the Tracker model allows it to be blank... I will make it mandatory for filling a BTS tracker.

costaconrado commented on apps/trackers/tests/test_service.py:
```diff
@@ -0,0 +1,110 @@
+"""
+    tests of Jira Task Manager service (Taskman)
```
Fixed, thank you.

costaconrado commented on apps/trackers/exceptions.py:
```diff
@@ -0,0 +1,11 @@
+"""
+OSIM exceptions
```
Fixed, thank you.

costaconrado commented on apps/trackers/service.py:
```diff
@@ -0,0 +1,105 @@
+"""
+Task Manager API endpoints
```
Fixed, thank you.

costaconrado commented on apps/trackers/bts_tracker.py:
```diff
@@ -0,0 +1,157 @@
+        description += f"OSIDB Flaw: {self._flaw.uuid}.\n"
+
+        if self._stream and self._stream.rhsa_sla_applicable:
+            description += f"{self._generate_sla()}\n"
```
Makes sense. Removed any SLA mention in this PR and left to be solved during OSIDB-96.

costaconrado commented on apps/trackers/bts_tracker.py:
```diff
@@ -0,0 +1,141 @@
+        """
+        Generates a text description for the bug being tracked
+        """
+        if not self._ps_module.private_trackers_allowed:
```
changed for actual embargoed status of the flaw instead of `PsModule.private_trackers_allowed`

costaconrado commented on apps/trackers/bts_tracker.py:
```diff
@@ -0,0 +1,141 @@
+        else:
+            description = "Security Tracking Issue\n\nDo not make this issue public.\n"
+
+        description += f"OSIDB Flaw: {self._flaw.uuid}.\n"
```
Added the url instead of the uuid.

costaconrado commented on apps/trackers/bts_tracker.py:
```diff
@@ -0,0 +1,157 @@
+            bts_tracker["fields"]["security"] = {"name": "Embargoed Security Issue"}
+        stream_name = "" if not self._stream else self._stream.name
+        existing_tracker = self._affect.trackers.filter(
+            type=Tracker.TrackerType.JIRA, ps_update_stream=stream_name
```
You are right, type would not be necessary here since affect will always have the same component from the same BTS. About the second comment, I believe it is already filtering by the affect itself (`self._affect.trackers`), is that right? I have removed the type, please take a look if now it is correct.

osoukup replied:
You are right that this is already narrowed to a single affect. Therefore the stream is the unique identifier.

costaconrado commented on collectors/jiraffe/tests/test_collectors.py:
```diff
@@ -149,3 +156,27 @@ def test_collect_id_known_linked(self):
+        """
+        Test that collector are able ot get metadata from Jira projects
+        """
+        if HTTPS_PROXY:
```
makes sense, duplicated `pin_urls` method from taskman conftest file, since it is in different modules it would not autouse here.

costaconrado replied:
the conftest with this fixture is from another module that has no affect on this class, if preferable I can add this fixture for this collector tests, but will also impact the read-only tests that uses production cluster.

osoukup replied:
You are right, I confused the two modules while they are clearly different. I am leaving this up to you. I would prefer the test environment to be stable/constant where it is possible/makes sense and not conditional.

costaconrado commented on apps/trackers/constants.py:
```diff
@@ -0,0 +1,7 @@
+TRACKER_API_VERSION = "v1"
+
+JIRA_SERVER = get_env("JIRA_URL", default="https://issues.redhat.com")
+JIRA_TEST_SERVER = get_env("JIRA_TEST_URL")
```
removed `JIRA_TEST_SERVER` and `JIRA_SERVER` variable since they are only used during tests and added them in  `conftest.py`

costaconrado commented on docker-compose.test.yml:
```diff
@@ -45,6 +45,7 @@ services:
         BZIMPORT_BZ_URL: ${BZIMPORT_BZ_URL}
         JIRA_AUTH_TOKEN: ${JIRA_AUTH_TOKEN:?Variable JIRA_AUTH_TOKEN must be set.}
         JIRA_URL: ${JIRA_URL}
+        JIRA_TEST_URL: ${JIRA_TEST_URL}
```
removed `JIRA_TEST_SERVER` and `JIRA_SERVER` variable since they are only used during tests and added them in  `conftest.py`

costaconrado commented on docs/developer/DEVELOP.md:
```diff
@@ -23,6 +23,10 @@ JIRA_AUTH_TOKEN=####################
 
+# Jira URL used for tests that requires writing trackers (optional)
+# Used only when capturing new VCRs
+JIRA_TEST_URL="https://foo.bar"
```
removed `JIRA_TEST_SERVER` and `JIRA_SERVER` variable since they are only used during tests and added them in  `conftest.py`

costaconrado commented:
@osoukup  I have addressed individually the comments but decided to also summarize the changes I made to make review less painful :sweat_smile:.

`apps/trackers/bts_tracker.py`
- Fixed class description
- Removed Jira priorities not contemplated by OJA-PRIS-001
- Removed any SLA mention and left to be solved during OSIDB-96
- Changed requested methods (`_generate_summary` and `_generate_description`), from JiraTracker to BTSTracker for reuse purpose in Bugzilla
- In `_generate_summary` method, considered the actual embargoed status of the flaw instead of `PsModule.private_trackers_allowed`
- Added the url for the flaw endpoint instead of `uuid`
- during the BTS object generation (`generate_bts_object`) I have removed the filter by tracker type since it would do nothing.

`apps/trackers/apps.py`
- Fixed app description

`collectors/jiraffe/collectors.py`
- removed hardcoded projects
- filtered projects that does not contains active PS update streams and projects that is not supported anymore
- Generate error logs in case collector fail to fetch Jira project metadata 

`collectors/jiraffe/tasks.py`
- Changed collector cronjob for 2:35 that is a time window not used by other collectors

`apps/trackers/service.py`
- changed validation during the tracker creation in order to limit 1 tracker per stream for each affect

`apps/trackers/constants.py`
`docs/developer/DEVELOP.md`
`docker-compose.test.yml`
`.github/workflows/test.yml`
- removed `JIRA_TEST_SERVER` variable since it is only used during tests and added it in the `conftest.py`

`apps/trackers/tests/conftest.py`
- duplicated `pin_urls` method from taskman's conftest file, since it is in different modules it would not autouse here.

osoukup reviewed (approved):
LGTM. Great job!
